cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tether/TetherClientHandler.java

```diff
@@ -0,0 +1,119 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.tether;
+
+import com.google.gson.Gson;
+import io.cdap.cdap.common.BadRequestException;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.internal.remote.RemoteAuthenticator;
+import io.cdap.common.http.HttpMethod;
+import io.cdap.common.http.HttpResponse;
+import io.cdap.http.AbstractHttpHandler;
+import io.cdap.http.HandlerContext;
+import io.cdap.http.HttpResponder;
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Optional;
+import javax.inject.Inject;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+
+/**
+ * {@link io.cdap.http.HttpHandler} to manage tethering client v3 REST APIs
+ */
+@Path(Constants.Gateway.API_VERSION_3)
+public class TetherClientHandler extends AbstractHttpHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(TetherClientHandler.class);
+  private static final Gson GSON = new Gson();
+  static final String CREATE_TETHER = "/v3/tethering/connect";
+
+  private final TetherStore store;
+  private final CConfiguration cConf;
+  private final String instanceName;
+
+  @Inject
+  TetherClientHandler(CConfiguration cConf, TetherStore store) {
+    this.store = store;
+    this.cConf = cConf;
+    this.instanceName = cConf.get(Constants.INSTANCE_NAME);
+  }
+
+  @Override
+  public void init(HandlerContext context) {
+    super.init(context);
+    Class<? extends RemoteAuthenticator> authClass = cConf.getClass(Constants.Tether.CLIENT_AUTHENTICATOR_CLASS,
+                                                                    null,
+                                                                    RemoteAuthenticator.class);
+    if (authClass != null) {
+      try {
+        RemoteAuthenticator.setDefaultAuthenticator(authClass.newInstance());
+      } catch (Exception e) {
+        LOG.error("Failed to set default authenticator", e);
+      }
+    }
+  }
+
+  /**
+   * Initiates tethering with the server.
+   */
+  @POST
+  @Path("/tethering/create")
+  public void createTether(FullHttpRequest request, HttpResponder responder) throws Exception {
+    String content = request.content().toString(StandardCharsets.UTF_8);
+    TetherCreationRequest tetherCreationRequest = GSON.fromJson(content, TetherCreationRequest.class);
+
+    List<PeerInfo> peers = store.getPeers();
+    Optional<PeerInfo> peer = peers.stream()
+      .filter(p -> p.getName().equals(tetherCreationRequest.getPeer()))
```

Comment:
Why get all peers and only look for one? Should just do a get call?

Reply:
fixed